node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -478,82 +483,268 @@ mod tests {
+                .into_iter()
+                .map(|vidx| {
+                    let secret = TestSecret(vidx.0);
+                    let (av, effects) = ActiveValidator::new(vidx, secret, start_time, &state);
+                    if let [Eff::ScheduleTimer(timestamp)] = *effects {
+                        // We start at time 410, with round length 16, so the first leader tick is
+                        // 416, and the first witness tick 426.
+                        if state.leader(earliest_round_start) == vidx {
+                            assert_eq!(
+                                timestamp,
+                                416.into(),
+                                "Invalid initial timer scheduled for {:?}.",
+                                vidx
+                            )
+                        } else {
+                            let witness_offset = av
+                                .witness_offset(state::round_len(state.params().init_round_exp()));
+                            let witness_timestamp = earliest_round_start + witness_offset;
+                            assert_eq!(
+                                timestamp, witness_timestamp,
+                                "Invalid initial timer scheduled for {:?}.",
+                                vidx
+                            )
+                        }
+                        timers.insert((timestamp, vidx));
+                    } else {
+                        panic!("Only one initial effect expected")
+                    }
+                    (vidx, av)
+                })
+                .collect();
+
+            TestState {
+                rng,
+                state,
+                instance_id,
+                fd,
+                validators: validator_map,
+                timers,
+            }
+        }
+
+        /// Force the validator to handle timer that may not have been sheduled by it.
+        /// Useful for testing.
+        /// Returns effects created when handling the timer.
+        fn handle_timer(
+            &mut self,
+            validator: ValidatorIndex,
+            timestamp: Timestamp,
+        ) -> Vec<Effect<TestContext>> {
+            // Remove the timer from the queue if it has been scheduled.
+            let _ = self.timers.remove(&(timestamp, validator));
+            let effects = self.validators.get_mut(&validator).unwrap().handle_timer(
+                timestamp,
+                &self.state,
+                self.instance_id,
+                &mut self.rng,
+            );
+            self.schedule_timer(validator, &effects);
+            self.add_new_unit(&effects);
+            effects
+        }
+
+        /// Propose new consensus value as validator `vidx`.
+        /// Returns effects created when proposing and newly proposed wire unit.
+        fn propose(
+            &mut self,
+            vidx: ValidatorIndex,
+            cv: <TestContext as Context>::ConsensusValue,
```

Comment:
I'm always confused about when this is or isn't allowed:
```suggestion
            cv: TestContext::ConsensusValue,
```

Reply:
I'm getting `ambiguous associated type` error.